internal/system/agent/router.go

```diff
@@ -40,56 +42,52 @@ func LoadRestRoutes() *mux.Router {
 }
 
 func operationHandler(w http.ResponseWriter, r *http.Request) {
-	if r.Body != nil {
-		defer r.Body.Close()
-	}
-
-	// TODO: Work with parsing mux.Vars(r) and assigning to vars.
-	var services []string
-	vars := mux.Vars(r)
-	action := vars["action"]
-
-	var params []string
-	var o models.Operation
-	dec := json.NewDecoder(r.Body)
-	err := dec.Decode(&o)
+	defer r.Body.Close()
 
+	b, err := ioutil.ReadAll(r.Body)
```

Comment:
Personal nitpick, it's more efficient and safer to just do as originally implemented like:

```go
o := models.Operation{}
err := json.NewDecoder(r.Body).Decode(&o)
// other stuff here
```

This has a few advantages:
1. If someone crafts a bad request like `"{} ... 20 GB of worthless random binary data follows"`, with ioutil.ReadAll you will end up attempting to read all of that into memory and have problems.
2. For the same reason, you end up duplicating the memory needed because you'll have one copy of it somewhere in the body stored internally behind `r.Body`, and then you make a copy of it inside `b` here, whereas with the `json.NewDecoder`, it doesn't actually perform any decoding until you call `Decode`, and just puts what it finds exactly into the object referred to by `&o`. 

That said, what you have works fine so I'll approve for Delhi, but just in general using `json.NewDecoder` is safer for decoding HTTP requests.

Reply:
I see your point about the large request and the memory utilization. However, I do like this approach because it provides a way through the Unmarshaler interface to initialize a type using custom business logic if necessary. It isn't just a deserialization handler but a more full-featured hook. I'll have to look into it further.

Reply:
I'm pretty sure that if the obiect implements the Unmarshaler interface that method will still be used, but more efficiently, but I haven't tested this in a while so I may be wrong on that point.